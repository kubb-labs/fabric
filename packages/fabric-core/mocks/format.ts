import type { Options } from 'prettier'
import { format as prettierFormat } from 'prettier'
import pluginTypescript from 'prettier/plugins/typescript'

const formatOptions: Options = {
  tabWidth: 2,
  printWidth: 160,
  parser: 'typescript',
  singleQuote: true,
  semi: false,
  bracketSameLine: false,
  endOfLine: 'auto',
  plugins: [pluginTypescript],
}
export function format(source?: string): Promise<string> {
  if (!source) {
    return Promise.resolve('')
  }

  try {
    return prettierFormat(source, formatOptions)
  } catch (_e) {
    return Promise.resolve(source)
  }
}
