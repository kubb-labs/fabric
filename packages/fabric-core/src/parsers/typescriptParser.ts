import path from 'node:path'
import ts from 'typescript'
import { getRelativePath } from '../utils/getRelativePath.ts'
import { trimExtName } from '../utils/trimExtName.ts'
import { createParser } from './createParser.ts'

const { factory } = ts

type PrintOptions = {
  source?: string
  baseName?: string
  scriptKind?: ts.ScriptKind
}

/**
 * Escaped new lines in code with block comments so they can be restored by {@link restoreNewLines}
 */
const escapeNewLines = (code: string) => code.replace(/\n\n/g, '\n/* :newline: */')

/**
 * Reverses {@link escapeNewLines} and restores new lines
 */
const restoreNewLines = (code: string) => code.replace(/\/\* :newline: \*\//g, '\n')

/**
 * Convert AST TypeScript/TSX nodes to a string based on the TypeScript printer.
 * Ensures consistent output across environments.
 * Also works as a formatter when `source` is provided without `elements`.
 */
export function print(elements: Array<ts.Node> = [], { source = '', baseName = 'print.tsx', scriptKind = ts.ScriptKind.TSX }: PrintOptions = {}): string {
  const sourceFile = ts.createSourceFile(baseName, escapeNewLines(source), ts.ScriptTarget.ES2022, true, scriptKind)

  const printer = ts.createPrinter({
    omitTrailingSemicolon: true,
    newLine: ts.NewLineKind.LineFeed,
    removeComments: false,
    noEmitHelpers: true,
  })

  let output: string

  if (elements.length > 0) {
    // Print only provided nodes
    const nodes = elements.filter(Boolean).sort((a, b) => (a.pos ?? 0) - (b.pos ?? 0))
    output = printer.printList(ts.ListFormat.MultiLine, factory.createNodeArray(nodes), sourceFile)
  } else {
    // Format the whole file
    output = printer.printFile(sourceFile)
  }

  return restoreNewLines(output).replace(/\r\n/g, '\n')
}

export function createImport({
  name,
  path,
  root,
  isTypeOnly = false,
  isNameSpace = false,
}: {
  name: string | Array<string | { propertyName: string; name?: string }>
  path: string
  root?: string
  isTypeOnly?: boolean
  isNameSpace?: boolean
}) {
  const resolvePath = root ? getRelativePath(root, path) : path

  if (!Array.isArray(name)) {
    let importPropertyName: ts.Identifier | undefined = factory.createIdentifier(name)
    let importName: ts.NamedImportBindings | undefined

    if (isNameSpace) {
      importPropertyName = undefined
      importName = factory.createNamespaceImport(factory.createIdentifier(name))
    }

    return factory.createImportDeclaration(
      undefined,
      factory.createImportClause(isTypeOnly, importPropertyName, importName),
      factory.createStringLiteral(resolvePath),
      undefined,
    )
  }

  return factory.createImportDeclaration(
    undefined,
    factory.createImportClause(
      isTypeOnly,
      undefined,
      factory.createNamedImports(
        name.map((item) => {
          if (typeof item === 'object') {
            const obj = item as { propertyName: string; name?: string }
            if (obj.name) {
              return factory.createImportSpecifier(false, factory.createIdentifier(obj.propertyName), factory.createIdentifier(obj.name))
            }

            return factory.createImportSpecifier(false, undefined, factory.createIdentifier(obj.propertyName))
          }

          return factory.createImportSpecifier(false, undefined, factory.createIdentifier(item))
        }),
      ),
    ),
    factory.createStringLiteral(resolvePath),
    undefined,
  )
}

export function createExport({
  path,
  asAlias,
  isTypeOnly = false,
  name,
}: {
  path: string
  asAlias?: boolean
  isTypeOnly?: boolean
  name?: string | Array<ts.Identifier | string>
}) {
  if (name && !Array.isArray(name) && !asAlias) {
    console.warn(`When using name as string, asAlias should be true ${name}`)
  }

  if (!Array.isArray(name)) {
    const parsedName = name?.match(/^\d/) ? `_${name?.slice(1)}` : name

    return factory.createExportDeclaration(
      undefined,
      isTypeOnly,
      asAlias && parsedName ? factory.createNamespaceExport(factory.createIdentifier(parsedName)) : undefined,
      factory.createStringLiteral(path),
      undefined,
    )
  }

  return factory.createExportDeclaration(
    undefined,
    isTypeOnly,
    factory.createNamedExports(
      name.map((propertyName) => {
        return factory.createExportSpecifier(false, undefined, typeof propertyName === 'string' ? factory.createIdentifier(propertyName) : propertyName)
      }),
    ),
    factory.createStringLiteral(path),
    undefined,
  )
}

export const typescriptParser = createParser({
  name: 'typescript',
  extNames: ['.ts', '.js'],
  install() {},
  async parse(file, options = { extname: '.ts' }) {
    const source = file.sources.map((item) => item.value).join('\n\n')

    const importNodes = file.imports
      .map((item) => {
        const importPath = item.root ? getRelativePath(item.root, item.path) : item.path
        const hasExtname = !!path.extname(importPath)

        return createImport({
          name: item.name,
          path: options.extname && hasExtname ? `${trimExtName(importPath)}${options.extname}` : item.root ? trimExtName(importPath) : importPath,
          isTypeOnly: item.isTypeOnly,
        })
      })
      .filter(Boolean)

    const exportNodes = file.exports
      .map((item) => {
        const exportPath = item.path

        const hasExtname = !!path.extname(exportPath)

        return createExport({
          name: item.name,
          path: options.extname && hasExtname ? `${trimExtName(item.path)}${options.extname}` : trimExtName(item.path),
          isTypeOnly: item.isTypeOnly,
          asAlias: item.asAlias,
        })
      })
      .filter(Boolean)

    return [file.banner, print([...importNodes, ...exportNodes]), source, file.footer].join('\n')
  },
})
