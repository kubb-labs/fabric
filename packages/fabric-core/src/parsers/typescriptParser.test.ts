import type ts from 'typescript'
import { format } from '../../mocks/format.ts'

import { print, createImport, createExport, typescriptParser } from './typescriptParser.ts'

const formatTS = (elements: ts.Node | (ts.Node | undefined)[]) => {
  return format(print([elements].flat().filter(Boolean)))
}

describe('TypeScript parser', () => {
  test('createImport', async () => {
    expect(
      await formatTS(
        createImport({
          name: 'hello',
          path: './hello.ts',
        }),
      ),
    ).toMatchSnapshot()

    expect(
      await formatTS(
        createImport({
          name: 'hello',
          path: './hello.ts',
          isTypeOnly: true,
        }),
      ),
    ).toMatchSnapshot()

    expect(
      await formatTS(
        createImport({
          name: ['hello'],
          path: './hello.ts',
        }),
      ),
    ).toMatchSnapshot()

    expect(
      await formatTS(
        createImport({
          name: 'hello',
          path: './hello.ts',
          isNameSpace: true,
        }),
      ),
    ).toMatchSnapshot()

    expect(
      await formatTS(
        createImport({
          name: [{ propertyName: 'hello', name: 'helloWorld' }],
          path: './hello.ts',
        }),
      ),
    ).toMatchSnapshot()
  })

  test('createExport', async () => {
    expect(
      await formatTS(
        createExport({
          path: './hello.ts',
        }),
      ),
    ).toMatchSnapshot()

    expect(
      await formatTS(
        createExport({
          name: ['hello', 'world'],
          asAlias: true,
          path: './hello.ts',
        }),
      ),
    ).toMatchSnapshot()

    expect(
      await formatTS(
        createExport({
          name: 'hello',
          asAlias: true,
          path: './hello.ts',
        }),
      ),
    ).toMatchSnapshot()

    try {
      await formatTS(
        createExport({
          name: 'hello',
          path: './hello.ts',
        }),
      )
    } catch (e) {
      expect(e).toBeDefined()
    }
  })

  test('typescriptParser.print combines banner/imports/exports/sources/footer and respects extname', async () => {
    const file = {
      path: '/project/src/index.ts',
      extname: '.ts',
      banner: '// banner',
      footer: '// footer',
      sources: [{ value: 'export const x = 1' }, { value: 'export const y = 2' }],
      imports: [
        { name: 'foo', path: './utils.ts' },
        { name: ['bar'], path: '/project/src/bar.js', root: '/project/src' },
      ],
      exports: [{ path: './hello.js' }, { name: ['alpha', 'beta'], path: './names.ts', asAlias: true }],
      meta: {},
    } as any

    const output = await typescriptParser.parse(file, { extname: '.ts' as any })
    expect(await format(output)).toMatchSnapshot()
  })
})
