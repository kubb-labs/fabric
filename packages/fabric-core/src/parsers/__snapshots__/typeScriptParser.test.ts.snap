// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TypeScript parser > createExport 1`] = `
"export * from './hello.ts'
"
`;

exports[`TypeScript parser > createExport 2`] = `
"export { hello, world } from './hello.ts'
"
`;

exports[`TypeScript parser > createExport 3`] = `
"export * as hello from './hello.ts'
"
`;

exports[`TypeScript parser > createImport 1`] = `
"import hello from './hello.ts'
"
`;

exports[`TypeScript parser > createImport 2`] = `
"import type hello from './hello.ts'
"
`;

exports[`TypeScript parser > createImport 3`] = `
"import { hello } from './hello.ts'
"
`;

exports[`TypeScript parser > createImport 4`] = `
"import * as hello from './hello.ts'
"
`;

exports[`TypeScript parser > createImport 5`] = `
"import { hello as helloWorld } from './hello.ts'
"
`;

exports[`TypeScript parser > typeScriptParser.print combines banner/imports/exports/sources/footer and respects extname 1`] = `
"// banner
import foo from './utils.ts'
import { bar } from './bar.ts'
export * from './hello.ts'
export { alpha, beta } from './names.ts'

export const x = 1

export const y = 2
// footer
"
`;

exports[`TypeScript parser > typescriptParser.print combines banner/imports/exports/sources/footer and respects extname 1`] = `
"// banner
import foo from './utils.ts'
import { bar } from './bar.ts'
export * from './hello.ts'
export { alpha, beta } from './names.ts'

export const x = 1

export const y = 2
// footer
"
`;
